Implement API to find the frame for a given virtual address

wipawel opened the pull request:
This implements #202.

wipawel committed:
mm/pmm: add frame finder functions for MFN and paddr

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
arch,pt: add frame finder for virtual address functionality

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

sktt commented on arch/x86/pagetables.c:
```diff
@@ -272,6 +273,80 @@ static void map_tmp_mapping_entry(void) {
+    _order = PAGE_ORDER_4K;
+
+done:
+    if (mfn)
+        *mfn = _mfn;
+    if (order)
+        *order = _order;
```
is it by design to allow null ptrs for mfn and order? Otherwise maybe possible to crash or warn as early as possible if it happens?

wipawel replied:
Main point was to allow either mfn or order pointers or both. We can add an ASSERT when there is none.

sktt replied:
Ok. For me fine as-is then

sktt commented on mm/pmm.c:
```diff
@@ -421,6 +421,83 @@ static frame_t *find_mfn_frame(list_head_t *list, mfn_t mfn, unsigned int order)
+frame_t *find_mfn_frame(mfn_t mfn, unsigned int order) {
+    frame_t *frame;
+
+    spin_lock(&lock);
+    frame = _find_mfn_frame(busy_frames, mfn, order);
+    if (!frame)
+        frame = _find_mfn_frame(free_frames, mfn, order);
+    spin_unlock(&lock);
```
Could we re-use `find_{busy,free}_mfn_frame` here? Or is it a bad idea with the lock-unlock-lock-unlock behavior imposed?

wipawel replied:
Yeah, it was done like that to avoid locking-unlocking problems. We now will have a static function to be used internally by PMM without locking and a bunch of public interfaces which handle the locking.

sktt commented on mm/pmm.c:
```diff
@@ -421,6 +421,83 @@ static frame_t *find_mfn_frame(list_head_t *list, mfn_t mfn, unsigned int order)
+frame_t *find_paddr_frame(paddr_t paddr) {
+    frame_t *frame;
+
+    spin_lock(&lock);
+    frame = _find_paddr_frame(busy_frames, paddr);
+    if (!frame)
+        frame = _find_paddr_frame(free_frames, paddr);
+    spin_unlock(&lock);
```
Same here but using `find_{busy,free}_paddr_frame` ?

sktt commented on mm/pmm.c:
```diff
@@ -491,13 +568,13 @@ static void merge_frames(frame_t *first) {
     if (FIRST_FRAME_SIBLING(first->mfn, first->order + 1)) {
         mfn_t next_mfn = NEXT_MFN(first->mfn, first->order);
-        second = find_mfn_frame(free_frames, next_mfn, first->order);
+        second = _find_mfn_frame(free_frames, next_mfn, first->order);
```
`find_free_mfn_frame` ?

sktt commented on mm/pmm.c:
```diff
@@ -491,13 +568,13 @@ static void merge_frames(frame_t *first) {
         mfn_t prev_mfn = PREV_MFN(first->mfn, first->order);
         second = first;
-        first = find_mfn_frame(free_frames, prev_mfn, first->order);
+        first = _find_mfn_frame(free_frames, prev_mfn, first->order);
```
`find_free_mfn_frame` ?

sktt commented on mm/pmm.c:
```diff
@@ -547,7 +624,7 @@ void put_free_frames(mfn_t mfn, unsigned int order) {
 
     spin_lock(&lock);
-    frame = find_mfn_frame(busy_frames, mfn, order);
+    frame = _find_mfn_frame(busy_frames, mfn, order);
```
`find_busy_mfn_frame` ?

sktt commented on include/mm/pmm.h:
```diff
@@ -76,6 +76,13 @@ extern frame_t *get_free_frames(unsigned int order);
 extern void put_free_frames(mfn_t mfn, unsigned int order);
 extern void reclaim_frame(mfn_t mfn, unsigned int order);
 
+extern frame_t *find_free_mfn_frame(mfn_t mfn, unsigned int order);
+extern frame_t *find_busy_mfn_frame(mfn_t mfn, unsigned int order);
+extern frame_t *find_mfn_frame(mfn_t mfn, unsigned int order);
+extern frame_t *find_free_paddr_frame(paddr_t paddr);
+extern frame_t *find_busy_paddr_frame(paddr_t paddr);
+extern frame_t *find_paddr_frame(paddr_t paddr);
```
I think either we should use the free,busy sugar functions, or just not include them at all?

wipawel replied:
For now, with the current PMM implementation, the split for free/busy and any (`find_mfn_frame`) is a small optimization as main current needs and uses are very specific (and for busy frames), but general API has been created too.

sktt replied:
Okey. I think I didn't think of them as API for tests that we write.

This is pretty nice, that we can allocate guest-physical memory easily. And the host physical base mapping can actually a small qemu wrapper manage :D

sktt commented on mm/pmm.c:
```diff
@@ -491,13 +568,13 @@ static void merge_frames(frame_t *first) {
 
```
Alternatively, don't we need to take the lock here?

wipawel replied:
All `static` functions are considered internal to the PMM implementation. It is better to take locks as externally as possible (external interfaces, potentially shared by different code other than PMM) to avoid locking/unlocking mishaps. Hence the extra interfaces to handle locking consistently. The `merge_frames()` and `split_frames()` are considered internal and assumed to operate with a taken lock.

sktt reviewed (commented):
I resolved a few comments. One question mainly remains

wipawel committed:
arch,pt: add get_va_mfn_order() functionality

The functions traverse specified page tables and obtain a MFN and frame
order for a specified virtual address.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

minipli-oss commented on include/mm/pmm.h:
```diff
@@ -100,6 +107,16 @@ static inline frame_t *get_first_frame(list_head_t *frames, unsigned int order)
+        return false;
+
+    paddr_t start_pa = mfn_to_paddr(frame->mfn);
+    paddr_t end_pa = start_pa + ORDER_TO_SIZE(frame->order);
+
+    return pa >= start_pa && pa < end_pa;
```
Should this be below instead to handle the very last pa (and unsigned overflow) as well?:
```c
    paddr_t end_pa = start_pa + ORDER_TO_SIZE(frame->order) - 1;

    return pa >= start_pa && pa <= end_pa;
```

wipawel replied:
Fixed. Thanks.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -446,6 +446,103 @@ int vunmap_user(void *va, mfn_t *mfn, unsigned int *order) {
+        goto done;
+    }
+
+    if (mfn_invalid(l3e->mfn) || !l3e->P)
+        return -ENOENT;
```
Doesn't this test need to come first, i.e. before the `l3e->PS` one?

wipawel replied:
Fixed. Thanks.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -446,6 +446,103 @@ int vunmap_user(void *va, mfn_t *mfn, unsigned int *order) {
+        goto done;
+    }
+
+    if (mfn_invalid(l2e->mfn) || !l2e->P)
+        return -ENOENT;
```
Same here as well. If the present bit isn't set, the other bits have no meaning, according to the SDM.

wipawel replied:
Fixed. Thanks.

minipli-oss commented on arch/x86/pagetables.c:
```diff
@@ -446,6 +446,103 @@ int vunmap_user(void *va, mfn_t *mfn, unsigned int *order) {
+
+    tab = tmp_map_mfn(l2e->mfn);
+    pte_t *l1e = l1_table_entry((pte_t *) tab, va);
+    _mfn = l1e->mfn;
```
The `mfn_invalid(l1e->mfn) || !l1e->P)` tests are missing?

wipawel replied:
Fixed. Thanks.

minipli-oss reviewed (changes requested):
Looks good, only minor changes needed.
